### ISSUE TITLE
Error "Missing @Injectable for field" with @Resource annotation

### ISSUE THREAD

**Issue by AlexStasko** [jmockit/jmockit1#359]
I use JMockit 1.28 (tried 1.29)
Java 1.8.92

I work with `@Resource` annotation. But with `name` value my tests not working. Without it tests work fine.
Console output:
```
java.lang.IllegalStateException: Missing @Injectable for field "by.common.model.DummyDomain dummyDomain" in DummyService
```

Also if using `@Autowired` with `@Qualifier` tests work fine.
**Test case is here:** [jmockit-check.zip](https://github.com/jmockit/jmockit1/files/580619/jmockit-check.zip)

I use something like this as workaround:
```
    @BeforeMethod
    public void setUp() {
        dummyService = new DummyService();
        ReflectionTestUtils.setField(dummyServie, "dummyDomain", mockDummyDomain);
    }
```

Might be that is a bug, maybe I 'm doing something wrong. Thank you in advance and hope for your help.

Thank you,
Alex

**Comment by rliesenfeld** [comment]
The @Injectable field should match the name of the target @Resource ("test"). So, name it "test".

**Comment by susau** [comment]
Will try and respond but the question why using @Autowired with @Qualifier works?

**Comment by AlexStasko** [comment]
@rliesenfeld in my project I use names like "subname1.subname2.name" and in this case I cannot name field with dots.

**Comment by rliesenfeld** [comment]
The names need to match, otherwise it's not possible to support a scenario where multiple @Resource's of the same type are injected into a given object.

Name matching can be made more flexible, though. I am thinking of two possible solutions, when given a resource name composed of multiple parts separated by special characters such as "." (dot) or "-" (hyphen):

1. Use only the last name part. So, if the resource is named "part1.part2.last", then the @Injectable would need to be named "last".
2. Use the full resource name, but replacing separator characters with the underscore "_" character, which can be used in the name of the injectable: "@Injectable Xyz part1_part2_last".

Which solution would you prefer?

**Comment by susau** [comment]
I am tech lead of the project so let me answer.
Var N2 looks definitely better but my previous question is still valid I think.

How possible to support a scenario where multiple `@Autowied` with `@Qualifier` the same type are injected into a given object? Because as I mentioned if change in the test case `@Resource ("test")` to 
`@Autowired` 
`@Qualifier("test)`
(which is same thing from Spring prospective) it works. Might be if multiple `@Resource` but with different types we can use the same "`@Autowired` "  approach?

**Comment by rliesenfeld** [comment]
I see the same behavior with @Resource(name = "test") and with @Qualifier("test"), as it should be.
For example, the test below passes:

```
public class ExampleDITest {
    static class CUT1 {
        @Resource(name = "db1")
        DataSource ds;
    }

    static class CUT2 {
        @Autowired @Qualifier("db2")
        DataSource ds;
    }

    @Tested CUT1 cut1;
    @Tested CUT2 cut2;

    @Test
    public void injectNamedResources(@Injectable DataSource db1, @Injectable DataSource db2) {
        assertNotSame(db1, db2);
        assertSame(db1, cut1.ds);
        assertSame(db2, cut2.ds);
    }
}
```

**Comment by AlexStasko** [comment]
Thanks for your suggestions. But if we will use solution with underscore we will face with codestyle problems. In this case can we use camelcase solution? Something like this:
name = "part1.part2.name"
field name = "part1Part2Name"

**Comment by susau** [comment]
>>codestyle problems
Good catch, I didn't release that. Camelcase is definitely better

**Comment by rliesenfeld** [comment]
Ok, camel case can be supported too.

**Comment by susau** [comment]
Is there any chance for fast release with that fix? 
Thanks

**Comment by Siarhei-Yarkavy** [comment]
Sorry, could this new behavior be documented in javadoc and tutorial?

**Comment by pedorro** [comment]
Hi.  I just wanted to follow up with this conversation.  While the fix added to v1.30 to support this seems to be working as described, it feels to me that it's not a very obvious way to deal with this.  Actually naming the Injectable variable with a name that must match the String of the Resource seems like an awkward and unexpected solution.

What if `@Injectable` could take a string parameter of the resource name being injected?  This seems to me to be simpler, distinctly more obvious, and more aligned with the pattern used by `@Resource` and `@Qualifier` for identifying the named resource.  This would eliminate the awkward issue of converting 'dot-delimited' bean names to valid Java variable names.  Also, sometimes the bean name string is a publicly defined String constant somewhere, making it super-easy to reuse the same value in the `@Injectable` that's used in the `@Resource`.

Thoughts?

**Comment by Spookyguy** [comment]
Ran into the same problem yesterday and found the solution here. My problem is, that the string in my resource annotation ist "java:jboss/mail/my-mailer". So I don't know how to name the appropriate variable. I think pedorros suggestion is a good one.
